README.md
# PySparkAI - WebSocket Chatbot Client for 星火大模型

PySparkAI 是一个基于WebSocket调用星火大模型API的客户端。其目的是为用户提供高度封装的方法，从而简化使用过程。

## 特性

- 基于WebSocket的实时聊天：确保及时、持续的交互
- 星火大模型API：调用先进的AI技术来提供精确的回答
- 高度封装：减少配置和引导时间，即插即用

## 快速开始

### 安装

```bash
pip install xunfeisparkai
```

### 使用

查看demo.py脚本以了解如何使用PySparkAI。以下是demo.py的简短摘要:

```python
from pysparkai import PySparkAI

def get_credentials_from_user():
    """
    获取用户输入的敏感信息。
    """
    APP_ID = "此处填入你的APP_ID"
    API_KEY = "此处填入你的API_KEY"
    API_SECRET = "此处填入你的API_SECRET"
    SPARK_URL = "此处填入你的SPARK_URL"
    DOMAIN = "此处填入你的DOMAIN"
	
    return APP_ID, API_KEY, API_SECRET, SPARK_URL, DOMAIN

# 以上所有信息都可以在你的讯飞开放平台的控制台中找到

def main():
    APP_ID, API_KEY, API_SECRET, SPARK_URL, DOMAIN = get_credentials_from_user()
    ai = PySparkAI(app_id=APP_ID, api_key=API_KEY, api_secret=API_SECRET, spark_url=SPARK_URL, domain=DOMAIN)
    prompt_message = {"role": "system",
                      "content": "在一个遥远的星球上，有一位智者名为Zarnak，他对于人类世界有着深厚的了解。"}
    question_message = {"role": "user",
                        "content": "Zarnak, 你能告诉我地球上的人的性格为什么多种多样吗？"}
    format_specification_message = {
        "role": "system", "content": "请使用第三人称，以故事的方式回答。"}
    messages = [prompt_message, format_specification_message, question_message]
    completion = ai.chat(messages, temperature=0.3, max_tokens=200)
    formatted_answer = f"""
            Prompt: {prompt_message["content"]}
            格式要求: {format_specification_message["content"]}
            问题: {question_message["content"]}
            回答: {completion["choices"][0]["message"]}
            """
    print(formatted_answer)
    print("Token 使用情况:")
    for key, value in completion["usage"].items():
        print(f"{key}: {value}")


if __name__ == "__main__":
    main()

```

完整的demo.py脚本在仓库中可用。

## 开发

### 依赖

- `websocket-client`: 为项目提供WebSocket通信功能。

### 贡献

如果您想为该项目做出贡献，请确保遵循以下步骤：

1. Fork这个仓库: [PySparkAI on GitLab](https://jihulab.com/lichman0405/sparkai.git)。当然，这里的GitLab是大陆本土化的**JihuLab**。
2. 创建你的功能分支 (`git checkout -b feature/YourFeature`)
3. 提交你的更改 (`git commit -am 'Add some feature'`)
4. 推送到分支 (`git push origin feature/YourFeature`)
5. 在GitLab上创建一个新的Merge Request
6. 你也能够在[PySparkAI on Github](https://github.com/lichman0405/pysparkai.git)上找到它。

## 许可

该项目使用MIT许可。

## 联系

如有任何问题或建议，请联系 [shadow.li981@gmail.com](mailto:shadow.li981@gmail.com).

## 文档

更多关于 `pysparkai` 的详细文档，包括如何安装、使用和其他信息，请访问 [这里](./docs/index.md)。


CHANGELOG.md
# Changelog

All notable changes to the PySparkAI project will be documented in this file.

## [Unreleased]

## [1.0.0] - 2023-08-30
### Added
- Initial release of PySparkAI.


src/sparkai/auth_websocket_client.py
# Description: This file implements a wrapper for the AI Chatbot API.
# Author: Shibo Li, MiQroEra Inc
# Date: 2023-08-30
# Version: v0.1


import threading
import base64
import hashlib
import hmac
import json
from urllib.parse import urlparse, urlencode
from datetime import datetime
from time import mktime
from wsgiref.handlers import format_date_time
import websocket
from api_utils import gen_params
from error_codes import get_error_message


class ChatbotClient:
    """
    ChatbotClient is a wrapper for the AI Chatbot API.
    """

    def __init__(self, appid, api_key, api_secret, spark_url, domain):
        """
        :param appid: The appid of the AI Chatbot API.
        :param api_key: The api_key of the AI Chatbot API.
        :param api_secret: The api_secret of the AI Chatbot API.
        :param spark_url: The spark_url of the AI Chatbot API.
        :param domain: The domain of the AI Chatbot API.
        """
        self.appid = appid
        self.api_key = api_key
        self.api_secret = api_secret
        self.spark_url = spark_url
        self.domain = domain
        self.answer = ""
        self.ws_param = self.Ws_Param(appid, api_key, api_secret, spark_url)
        self.usage_questions_tokens = 0
        self.usage_prompt_tokens = 0
        self.usage_completion_tokens = 0
        self.usage_total = 0

    class Ws_Param:
        """
        Ws_Param is a wrapper for the websocket parameters.
        """

        def __init__(self, APPID, APIKey, APISecret, Spark_url):
            self.APPID = APPID
            self.APIKey = APIKey
            self.APISecret = APISecret
            self.host = urlparse(Spark_url).netloc
            self.path = urlparse(Spark_url).path
            self.Spark_url = Spark_url

        def create_url(self):
            """
            The method is to create the url for the websocket.
            :return: The url for the websocket.
            """
            now = datetime.now()
            date = format_date_time(mktime(now.timetuple()))

            signature_origin = f"host: {self.host}\ndate: {date}\nGET {self.path} HTTP/1.1"
            signature_sha = hmac.new(self.APISecret.encode('utf-8'), signature_origin.encode('utf-8'),
                                     digestmod=hashlib.sha256).digest()
            signature_sha_base64 = base64.b64encode(
                signature_sha).decode(encoding='utf-8')

            authorization_origin = f'api_key="{self.APIKey}", algorithm="hmac-sha256", headers="host date request-line", signature="{signature_sha_base64}"'
            authorization = base64.b64encode(
                authorization_origin.encode('utf-8')).decode(encoding='utf-8')

            v = {
                "authorization": authorization,
                "date": date,
                "host": self.host
            }
            url = self.Spark_url + '?' + urlencode(v)
            return url

    def on_message(self, ws, message):
        """
        The method is to handle the message received from the websocket.
        :param ws: The websocket object.
        :param message: The message received from the websocket.
        :return: None
        """
        data = json.loads(message)
        code = data['header']['code']
        if code != 0:
            error_message = get_error_message(code)
            print(f'请求错误: {code}, {error_message}')
            ws.close()
        else:
            choices = data["payload"]["choices"]
            status = choices["status"]
            content = choices["text"][0]["content"]
            self.answer += content
            if 'usage' in data['payload']:
                question_tokens = data['payload']['usage']['text']['question_tokens']
                prompt_tokens = data['payload']['usage']['text']['prompt_tokens']
                completion_tokens = data['payload']['usage']['text']['completion_tokens']
                total_tokens = data['payload']['usage']['text']['total_tokens']
                self.usage_questions_tokens += question_tokens
                self.usage_prompt_tokens += prompt_tokens
                self.usage_completion_tokens += completion_tokens
                self.usage_total += total_tokens
                if status == 2:
                    ws.close()

    def on_error(self, ws, error):
        print("### error:", error)

    @staticmethod
    def on_close(ws, one, two):
        # print("")
        pass

    def on_open(self, ws):
        threading.Thread(target=self.run, args=(ws,)).start()

    def run(self, ws):
        data = json.dumps(gen_params(
            appid=self.appid,
            domain=self.domain,
            question=ws.question,
            temperature=ws.temperature,
            max_tokens=ws.max_tokens
        )
        )
        ws.send(data)

    def get_response(self, question, temperature=0.5, max_tokens=2048):
        """
        The method is to get the response from the AI Chatbot API.
        :param question: The question to be asked.
        :param temperature: The temperature of the AI Chatbot API.
        :param max_tokens: The max_tokens of the AI Chatbot API.
        :return: The response from the AI Chatbot API.
        """
        self.answer = ""

        if not question.strip():
            print("请您输入有效内容。")
            return

        websocket.enableTrace(False)
        ws_url = self.ws_param.create_url()
        ws = websocket.WebSocketApp(
            ws_url,
            on_message=self.on_message,
            on_error=self.on_error,
            on_close=self.on_close
        )
        ws.appid = self.appid
        ws.domain = self.domain
        ws.question = question
        ws.temperature = temperature or 0.7  # default 0.7
        ws.max_tokens = max_tokens or 4096   # default 4096
        ws.on_open = self.on_open
        threading.Timer(10, ws.close).start()
        ws.run_forever()
        return self.answer

src/sparkai/sparkai.py
# Description: SparkAI class for chatbot
# Author: Shibo Li, MiQroEra Inc
# Date: 2023-08-30
# Version: v0.1


import os
from .auth_websocket_client import ChatbotClient


class SparkAI:
    """
    SparkAI class for chatbot
    """

    def __init__(self, api_key, api_secret, app_id, spark_url, domain):
        self.API_KEY = api_key or os.getenv("SPARK_AI_API_KEY")
        self.API_SECRET = api_secret or os.getenv("SPARK_AI_API_SECRET")
        self.APP_ID = app_id or os.getenv("SPARK_AI_APP_ID")
        self.SPARK_URL = spark_url or os.getenv("SPARK_AI_SPARK_URL")
        self.domain = domain or os.getenv("SPARK_AI_DOMAIN")

        # Initialize the client
        self.client = ChatbotClient(
            self.APP_ID, self.API_KEY, self.API_SECRET, self.SPARK_URL, self.domain)

    def chat(self, messages, temperature=None, max_tokens=None):
        """
        The method is to define a chatbot.
        :param messages: The messages of the chatbot.
        :param temperature: The temperature of the chatbot.
        :param max_tokens: The max_tokens of the chatbot.
        :return: The response of the chatbot.
        """
        user_input = messages[-1]["content"]
        response = self.client.get_response(
            user_input, temperature=temperature, max_tokens=max_tokens)
        response_message = {
            "message": response,
            "role": "assistant"
        }
        token_counts = {
            'question_tokens': self.client.usage_questions_tokens,
            'prompt_tokens': self.client.usage_prompt_tokens,
            'completion_tokens': self.client.usage_completion_tokens,
            'total_tokens': self.client.usage_total
        }
        return {
            "choices": [response_message],
            "usage": token_counts
        }